### PR TITLE
Support for chemistry questions

### PR DESCRIPTION
**Pull request by aw684**
Addition of chemical elements, state symbols and relations to support chemistry questions, and option to add particles, mass number and proton number for nuclear equations. Improved kerning, altered available docking points and made other bug fixes.

### CODE REVIEW

File: app/ts/inequality/Widget.ts [hunk 1]
```diff
@@ -35,64 +35,76 @@ class Rect {
...
 
-	/** Convenience pointer to this widget's parent */
-	parentWidget: Widget = null;
+    /** An array of the types of docking points that this widget can dock to */
+    docksTo: Array<string> = [];
 
-	isHighlighted = false;
-	color = null;
-	isMainExpression = false;
+    /** Convenience pointer to this widget's parent */
```
Comment (donderful) [inline app/ts/inequality/Widget.ts]: Test:
You mean convenient?

Comment (Morpheu5) [inline app/ts/inequality/Widget.ts]: No, I actually mean convenience, although it is convenient.

File: app/js/app/app.js [hunk 2]
```diff
@@ -89,7 +89,8 @@ define([
 
         // Here we configure the api provider with the server running the API. Don't need to do this if we want to use the same server as the static content.
         if (document.location.hostname == "localhost") {
-            apiProvider.urlPrefix("http://localhost:8080/isaac-api/api");
+            apiProvider.urlPrefix("https://staging.isaacphysics.org/api/any/api");
```
Comment (jsharkey13) [inline app/js/app/app.js]: These lines will need to be returned back to what they were, and you can just have the changes locally without committing them.

File: app/js/app/directives/equation_editor/EquationEditor.js [hunk 3]
```diff
@@ -3,47 +3,51 @@ define(function(require) {
...
-
                 scope.equationEditorElement = element;
 
                 scope.selectedSymbols = [];
                 scope.selectionHandleFlags = {
                     showCalc: false,
                     showResize: true,
-                    showMove: true
+                    showMove: false
                 };
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/EquationEditor.js]: The `selectionHandleFlags` are leftovers from a now-unused system, where if you clicked on a symbol you could use a menu to drag it round/change it to a vector/bold it/etc. Are we likely to have that back, and either way do we need these lines?

File: app/js/app/directives/equation_editor/EquationEditor.js [hunk 4]
```diff
@@ -3,47 +3,51 @@ define(function(require) {
...
-                    element.find(".top-menu").css({"bottom": scope.equationEditorElement.height()}).removeClass("active-menu");
+                    element.find(".top-menu").css({
+                        "bottom": scope.equationEditorElement.height()
+                    }).removeClass("active-menu");
                 });
 
                 scope.$on("newSymbolDrag", function(_, symbol, pageX, pageY, mousePageX, mousePageY) {
-                    scope.draggingNewSymbol = true;
 
+                    console.debug(scope.dimmensions);
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/EquationEditor.js]: This `console.debug(...)` needs to go. I hope dimensions was spelt correctly elsewhere? :P

File: app/js/app/directives/equation_editor/EquationEditor.js [hunk 5]
```diff
@@ -111,20 +117,41 @@ define(function(require) {
...
+
+                        if (editorMode == "maths" && questionDoc && questionDoc.availableSymbols) {
                             var parsed = parseCustomSymbols(questionDoc.availableSymbols);
                             if (parsed.vars.length > 0) {
                                 scope.symbolLibrary.customVars = parsed.vars;
                             }
                             if (parsed.fns.length > 0) {
                                 scope.symbolLibrary.customFunctions = parsed.fns;
                             }
+                            var parsedOps = parseCustomOperations(questionDoc.availableSymbols);
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/EquationEditor.js]: Is it not possible to make `parseCustomSymbols(...)`  also deal with operators too? That would be nice! make sure you avoid null pointers when returning `parsed.vars`, `parsed.fns` and the new `parsed.ops` though!

Do we need all these `console.dubug`s now?

File: app/js/app/directives/equation_editor/EquationEditor.js [hunk 6]
```diff
@@ -176,24 +210,32 @@ define(function(require) {
...
+                var greekLetters = ["\\alpha", "\\beta", "\\gamma", "\\delta", "\\varepsilon", "\\zeta", "\\eta", "\\theta", "\\iota", "\\kappa", "\\lambda", "\\mu", "\\nu", "\\xi", "\\omicron", "\\pi", "\\rho", "\\sigma", "\\tau", "\\upsilon", "\\phi", "\\chi", "\\psi", "\\omega"];
+                var greekLettersUpper = ["\\Gamma", "\\Delta", "\\Theta", "\\Lambda", "\\Xi", "\\Pi", "\\Sigma", "\\Upsilon", "\\Phi", "\\Psi", "\\Omega"];
+                var elements = ["H", "He", "Li", "Be", "B", "C", "N", "O", "F", "Ne", "Na", "Mg", "Al", "Si", "P", "S", "Cl", "Ar", "K", "Ca", "Sc", "Ti", "V", "Cr", "Mn", "Fe", "Co", "Ni", "Cu", "Zn", "Ga", "Ge", "As", "Se", "Br", "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Tc", "Ru", "Rh", "Pd", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I", "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu", "Hf", "Ta", "W", "Re", "Os", "Ir", "Pt", "Au", "Hg", "Tl", "Pb", "Bi", "Po", "At", "Rn", "Fr", "Ra", "Ac", "Th", "Pa", "U", "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr", "Rf", "Db", "Sg", "Bh", "Hs", "Mt", "Ds", "Rg", "Cn", "Uut", "Fl", "Uup", "Lv", "Uus", "Uuo"];
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/EquationEditor.js]: This is going to be fun. @donderful and @aw684 - make sure you both have the same list of elements. Take care with `Uub`, `Uuq`, `Uuh`: these should definitely be `Cn`, `Fl` and `Lv` now.

Also, by November the following symbols might have been approved:
`Uut`  ->  `Nh`
`Uup`  ->  `Mc`
`Uus`  ->  `Ts`
`Uuo`  ->  `Og`

File: app/js/app/directives/equation_editor/EquationEditor.js [hunk 7]
```diff
@@ -727,31 +1134,60 @@ define(function(require) {
...
-                        texLabel: true
+                                scope.calculusTitle = {
+                                    type: "string",
+                                    menu: {
+                                        label: "\\int",
+                                        texLabel: true
+                                    }
+                                };
+                */
+                var parseCustomOperations = function(symbols) {
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/EquationEditor.js]: Can we merge this into `parseCustomSymbols()`?

File: app/js/app/directives/equation_editor/MenuSymbol.js [hunk 8]
```diff
@@ -82,37 +93,63 @@ define([], function() {
 
                     element.css("left", 0);
                     element.css("top", 0);
-                    
-                    scope.$emit("symbolDrop", scope.symbol, pageX, pageY, pageX, pageY);
+
+                    // This ensures new symbols can be selected.
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/MenuSymbol.js]: What does the below code do, and why? What are the debugs for, and if they're staying - why the odd numbering?

File: app/js/app/directives/equation_editor/TopMenu.js [hunk 9]
```diff
@@ -23,7 +25,8 @@ define([], function() {
 
                 var el = element.find(".top-menu");
 
-                allMenus.push(el[0]);
+								allMenus.push(el[0]);
```
Comment (jsharkey13) [inline app/js/app/directives/equation_editor/TopMenu.js]: ???

File: app/js/app/router.js [hunk 10]
```diff
@@ -103,122 +107,144 @@ define(["angular-ui-router"], function() {
...
+            resolve: {
+                // BIG RED AND YELLOW WARNING WITH SPARKLES AND A FEW CRACKERS JUST IN CASE:
+                // we may want to revert this policy at some point.
+                // requireRole: getRolePromiseInjectableFunction(["ADMIN", "CONTENT_EDITOR", "EVENT_MANAGER"]),
+            },
+            views: {
+                "body": {
+                    templateUrl: "/partials/states/equation_editor.html",
+
+                    controller: ['$scope', '$rootScope', '$stateParams', function($scope, $rootScope, $stateParams) {
```
Comment (jsharkey13) [inline app/js/app/router.js]: Could we think about moving this to it's own file? It's getting a bit long to be inline now.

`EqualityPageController.js` perhaps, in with the others?

File: app/ts/inequality/ChemicalElement.ts [hunk 11]
```diff
@@ -0,0 +1,306 @@
+import { Widget, Rect } from './Widget.ts'
+import { BinaryOperation } from "./BinaryOperation.ts";
+import { DockingPoint } from "./DockingPoint.ts";
+import { Relation } from "./Relation.ts";
+import { Num } from "./Num.ts";
+import { StateSymbol } from "./StateSymbol.ts";
+
+/** A class for representing variables and constants (aka, elements). */
```
Comment (jsharkey13) [inline app/ts/inequality/ChemicalElement.ts]: You could just say a class for representing chemical elements . . .

File: app/ts/inequality/ChemicalElement.ts [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    /**
+     * Generates all the docking points in one go and stores them in this.dockingPoints.
+     * A ChemicalElement has three docking points:
+     *
+     * - _right_: Binary operation (addition, subtraction), ChemicalElement (multiplication)
+     * - _superscript_: Exponent
+     * - _subscript_: Subscript (duh?)
+     */
```
Comment (jsharkey13) [inline app/ts/inequality/ChemicalElement.ts]: This docstring needs updating.

File: app/ts/inequality/ChemicalElement.ts [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+     * Generates the expression corresponding to this widget and its subtree.
+     *
+     * The `subscript` format is a special one for generating symbols that will work with the sympy checker. It squashes
+     * everything together, ignoring operations and all that jazz.
+     *
+     * @param format A string to specify the output format. Supports: latex, python, subscript.
+     * @returns {string} The expression in the specified format.
+     */
+
+    // todo add mhchem with \alpha etc
```
Comment (jsharkey13) [inline app/ts/inequality/ChemicalElement.ts]: Done?

File: app/ts/inequality/ChemicalElement.ts [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+    getExpression(format: string): string {
+        var expression = "\\text{" + this.element + "}";
+
+        var isParticle = (this.element[0] != '\\');
+        if (format == "latex") {
+
+            expression = "\\text{" + this.element + "}";// need to remove this so that we can append the element to mass/proton numbers
+            // TODO: add support for mass/proton number, decide if we render both simultaneously or separately.
+            // Should we render one if the other is ommitted? - for now, no.
```
Comment (jsharkey13) [inline app/ts/inequality/ChemicalElement.ts]: Done? 
Always render them, whichever you have.

File: app/ts/inequality/ChemicalElement.ts [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+            }
+            if (this.dockingPoints["superscript"].child != null) {
+                expression += this.dockingPoints["superscript"].child.getExpression(format);
+            }
+            if (this.dockingPoints["right"].child != null) {
+                expression += this.dockingPoints["right"].child.getExpression(format);
+            }
+        } else if (format == "python") {
+            expression = "";
+        } else if (format == "mathml") {
```
Comment (jsharkey13) [inline app/ts/inequality/ChemicalElement.ts]: I don't necessarily like it; but I feel we should implement it at some point: https://www.drillster.com/info/mathml#reaction_equations

File: app/ts/inequality/Num.ts [hunk 16]
```diff
@@ -75,24 +69,40 @@ class Num extends Widget {
...
                     expression += this.dockingPoints["right"].child.getExpression(format);
                 } else {
                     // WARNING This assumes it's a Number, hence produces a multiplication
                     expression += this.dockingPoints["right"].child.getExpression(format);
                 }
+                console.debug("expression", expression);
             }
+        } else if (format == "mhchem") {
+            expression = this.getFullText("mhchem"); // need to remove this so that we can append the element to mass/proton numbers
+            // TODO: add support for mass/proton number, decide if we render both simultaneously or separately.
```
Comment (jsharkey13) [inline app/ts/inequality/Num.ts]: Is this relevant here? Has it been done?

File: app/ts/inequality/Particle.ts [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+        this.dockingPoints["right"] = new DockingPoint(this, this.p.createVector(box.w / 2 + this.s.mBox.w / 4, -this.s.xBox.h / 2), 1, "particle", "right");
+        this.dockingPoints["superscript"] = new DockingPoint(this, this.p.createVector(box.w / 2 + this.scale * 20, -this.scale * this.s.mBox.h), 0.666, "exponent", "superscript");
+        this.dockingPoints["subscript"] = new DockingPoint(this, this.p.createVector(box.w / 2 + this.scale * 20, descent), 0.666, "subscript", "subscript");
+        this.dockingPoints["mass_number"] = new DockingPoint(this, this.p.createVector(0, 0), 0.666, "top-left", "mass_number");
+        this.dockingPoints["proton_number"] = new DockingPoint(this, this.p.createVector(0, 0), 0.666, "bottom-left", "proton_number");
+    }
+
+    getExpression(format: string): string {
+        var expression = "";
+        console.debug("type:", this.type);
```
Comment (jsharkey13) [inline app/ts/inequality/Particle.ts]: Do we need these debugging lines any more?

File: app/ts/inequality/Relation.ts [hunk 18]
```diff
@@ -101,19 +120,25 @@ class Relation extends Widget {
      */
     getExpression(format: string): string {
         var expression = "";
-        if(format == "latex") {
+        console.debug("this", this);
```
Comment (jsharkey13) [inline app/ts/inequality/Relation.ts]: Do we need this?

File: app/ts/inequality/StateSymbol.ts [hunk 19]
```diff
@@ -0,0 +1,213 @@
+import { Widget, Rect } from './Widget.ts';
+import { Symbol } from './Symbol.ts';
+import { DockingPoint } from "./DockingPoint.ts";
+
+/**
+ * Relations, such as equalities, inequalities, and unexpected friends.
```
Comment (jsharkey13) [inline app/ts/inequality/StateSymbol.ts]: ???

State symbols!

File: app/ts/inequality/StateSymbol.ts [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+     * The `subscript` format is a special one for generating symbols that will work with the sympy checker. It squashes
+     * everything together, ignoring operations and all that jazz.
+     *
+     * @param format A string to specify the output format. Supports: latex, python, subscript.
+     * @returns {string} The expression in the specified format.
+     */
+    getExpression(format: string): string {
+        var expression = "";
+        if (format == "latex") {
+          expression = this.latexSymbol + "";
```
Comment (jsharkey13) [inline app/ts/inequality/StateSymbol.ts]: Why do we need the ` + "";` in all these cases? Surely if it's an implicit typecast, it'd be much nicer to do `expression += this.insertSymbolTypeHere` ?

File: app/ts/inequality/Relation.ts [hunk 21]
```diff
@@ -136,12 +161,12 @@ class Relation extends Widget {
         this.p.fill(this.color).strokeWeight(0).noStroke();
```
Comment (jsharkey13) [inline app/ts/inequality/Relation.ts]: The `token()` method (which this silly interface won't let me comment on) should return those symbols not in the menu by default. You may have to hard-code this list in two places and just hope that people update both; maybe note it in a comment if necessary, or else build support for ignoring the customSymbols which appear in the menu anyway and list all operators here?

File: app/ts/inequality/StateSymbol.ts [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+            if (dockingPoint.child != null) {
+                dockingPoint.child.scale = this.scale * dockingPoint.scale;
+                dockingPoint.child._shakeIt();
+                boxes[dockingPointName] = dockingPoint.child.boundingBox(); // NB: This only looks at the direct child!
+            }
+        });
+
+        // Calculate our own geometry
+
+        // Nothing to do for Relation
```
Comment (jsharkey13) [inline app/ts/inequality/StateSymbol.ts]: This isn't the Relation class any more!

**Pull request merged**